File: packages/public-api/src/GroupService.ts
import Fetch from "./Fetch.js";
import { handleResponse } from "./handleResponse.js";
import {
  Response,
  Problem,
  GroupResponse,
  GroupVersionsResponse,
  GroupVersionMembersResponse,
  GroupRestoreResponse,
  CreateGroupRequest,
  CloneGroupRequest,
  EditGroupRequest,
  AddListToGroupRequest,
  EditGroupModelRequest,
  AddGroupRerankerRequest,
  TransferGroupRequest,
} from "./Types.js";

const enc = encodeURIComponent;

export default class GroupService {
  private readonly path = `group`;
  constructor(private readonly fetch: Fetch) {}

  /** Creates a Group - search multiple Lists by combining them into a Group. */
  async create(request: CreateGroupRequest): Promise<Response<GroupResponse, Problem>> {
    const response = await this.fetch.post(this.path, request);
    return handleResponse(response, async () => (await response.json()) as GroupResponse);
  }

  /** Retrieves the details of a Group. */
  async get(account_name: string, group_name: string): Promise<Response<GroupResponse, Problem>> {
    const response = await this.fetch.get(`${this.path}/${enc(account_name)}/${enc(group_name)}`);
    return handleResponse(response, async () => (await response.json()) as GroupResponse);
  }

  /** Lists the membership versions of a Group, and which is current. */
  async versions(account_name: string, group_name: string): Promise<Response<GroupVersionsResponse, Problem>> {
    const response = await this.fetch.get(`${this.path}/versions/${enc(account_name)}/${enc(group_name)}`);
    return handleResponse(response, async () => (await response.json()) as GroupVersionsResponse);
  }

  /** Retrieves the member Lists pinned at a specific Group membership version. */
  async versionMembers(account_name: string, group_name: string, version: number): Promise<Response<GroupVersionMembersResponse, Problem>> {
    const response = await this.fetch.get(`${this.path}/versions/${enc(account_name)}/${enc(group_name)}/${version}`);
    return handleResponse(response, async () => (await response.json()) as GroupVersionMembersResponse);
  }

  /** Edits Group defaults (search/suggest size, units). */
  async edit(account_name: string, group_name: string, request: EditGroupRequest): Promise<Response<GroupResponse, Problem>> {
    const response = await this.fetch.put(`${this.path}/${enc(account_name)}/${enc(group_name)}`, request);
    return handleResponse(response, async () => (await response.json()) as GroupResponse);
  }

  /** Clones a Group into a new Group, copying its configuration and member Lists. */
  async clone(account_name: string, group_name: string, request: CloneGroupRequest): Promise<Response<GroupResponse, Problem>> {
    const response = await this.fetch.post(`${this.path}/clone/${enc(account_name)}/${enc(group_name)}`, request);
    return handleResponse(response, async () => (await response.json()) as GroupResponse);
  }

  /** Bumps the Group to a new membership version, pinning members to their latest versions. */
  async bumpVersion(account_name: string, group_name: string): Promise<Response<GroupResponse, Problem>> {
    const response = await this.fetch.post(`${this.path}/version/${enc(account_name)}/${enc(group_name)}`, undefined);
    return handleResponse(response, async () => (await response.json()) as GroupResponse);
  }

  /** Rolls the Group's membership forward to a past version. */
  async restore(account_name: string, group_name: string, version: number): Promise<Response<GroupRestoreResponse, Problem>> {
    const response = await this.fetch.post(`${this.path}/restore/${enc(account_name)}/${enc(group_name)}/${version}`, undefined);
    return handleResponse(response, async () => (await response.json()) as GroupRestoreResponse);
  }

  /** Adds a List (at a specific version) to the Group. */
  async addList(account_name: string, group_name: string, request: AddListToGroupRequest): Promise<Response<GroupResponse, Problem>> {
    const response = await this.fetch.put(`${this.path}/list/${enc(account_name)}/${enc(group_name)}`, request);
    return handleResponse(response, async () => (await response.json()) as GroupResponse);
  }

  /** Removes a List from the Group. */
  async removeList(account_name: string, group_name: string, list_name: string): Promise<Response<GroupResponse, Problem>> {
    const response = await this.fetch.delete(`${this.path}/list/${enc(account_name)}/${enc(group_name)}/${enc(list_name)}`);
    return handleResponse(response, async () => (await response.json()) as GroupResponse);
  }

  /** Sets the embedding model used by the Group (queued, converges asynchronously). */
  async setModel(account_name: string, group_name: string, request: EditGroupModelRequest): Promise<Response<GroupResponse, Problem>> {
    const response = await this.fetch.put(`${this.path}/model/${enc(account_name)}/${enc(group_name)}`, request);
    return handleResponse(response, async () => (await response.json()) as GroupResponse);
  }

  /** Removes the Group's explicit model, reverting to the default. */
  async removeModel(account_name: string, group_name: string): Promise<Response<GroupResponse, Problem>> {
    const response = await this.fetch.delete(`${this.path}/model/${enc(account_name)}/${enc(group_name)}`);
    return handleResponse(response, async () => (await response.json()) as GroupResponse);
  }

  /** Sets the reranker applied to Group search results. */
  async setReranker(account_name: string, group_name: string, request: AddGroupRerankerRequest): Promise<Response<GroupResponse, Problem>> {
    const response = await this.fetch.post(`${this.path}/reranker/${enc(account_name)}/${enc(group_name)}`, request);
    return handleResponse(response, async () => (await response.json()) as GroupResponse);
  }

  /** Removes the Group's reranker. */
  async removeReranker(account_name: string, group_name: string): Promise<Response<GroupResponse, Problem>> {
    const response = await this.fetch.delete(`${this.path}/reranker/${enc(account_name)}/${enc(group_name)}`);
    return handleResponse(response, async () => (await response.json()) as GroupResponse);
  }

  /** Moves the Group to a different subscription. */
  async transfer(account_name: string, group_name: string, request: TransferGroupRequest): Promise<Response<GroupResponse, Problem>> {
    const response = await this.fetch.put(`${this.path}/transfer/${enc(account_name)}/${enc(group_name)}`, request);
    return handleResponse(response, async () => (await response.json()) as GroupResponse);
  }

  /** Removes an existing Group. */
  async delete(account_name: string, group_name: string): Promise<Response<void, Problem>> {
    const response = await this.fetch.delete(`${this.path}/${enc(account_name)}/${enc(group_name)}`);
    return handleResponse(response, () => undefined);
  }
}

File: dotnet/SearchStack.PublicApi/Services/GroupService.cs
using System.Threading;
using System.Threading.Tasks;
using static SearchStack.PublicApi.Url;

namespace SearchStack.PublicApi;

/// <summary>Group create / delete operations.</summary>
public sealed class GroupService
{
    private const string Path = "group";
    private readonly Transport _transport;

    internal GroupService(Transport transport) => _transport = transport;

    /// <summary>Creates a Group - search multiple Lists by combining them into a Group.</summary>
    public Task<Response<GroupResponse>> CreateAsync(CreateGroupRequest request, CancellationToken cancellationToken = default) =>
        _transport.PostAsync<GroupResponse>(Path, request, cancellationToken);

    /// <summary>Clones a Group into a new Group, copying its configuration and member Lists. By default members are re-pinned to their latest version.</summary>
    public Task<Response<GroupResponse>> CloneAsync(string accountName, string groupName, CloneGroupRequest request, CancellationToken cancellationToken = default) =>
        _transport.PostAsync<GroupResponse>($"{Path}/clone/{Enc(accountName)}/{Enc(groupName)}", request, cancellationToken);

    /// <summary>Bumps a Group's version: re-pins every member List to its latest version and advances current_version. A no-op when every member is already current.</summary>
    public Task<Response<GroupResponse>> BumpVersionAsync(string accountName, string groupName, CancellationToken cancellationToken = default) =>
        _transport.PostAsync<GroupResponse>($"{Path}/version/{Enc(accountName)}/{Enc(groupName)}", new { }, cancellationToken);

    /// <summary>Restores a historical membership version as the Group's new current_version (roll-forward, not rewind). Member Lists deleted since that version are skipped and returned in skipped_lists.</summary>
    public Task<Response<GroupRestoreResponse>> RestoreVersionAsync(string accountName, string groupName, int version, CancellationToken cancellationToken = default) =>
        _transport.PostAsync<GroupRestoreResponse>($"{Path}/restore/{Enc(accountName)}/{Enc(groupName)}/{version}", new { }, cancellationToken);

    /// <summary>Lists every searchable membership version of a Group (pass one as group_version to search a frozen set of records).</summary>
    public Task<Response<GroupVersionsResponse>> GetVersionsAsync(string accountName, string groupName, CancellationToken cancellationToken = default) =>
        _transport.GetAsync<GroupVersionsResponse>($"{Path}/versions/{Enc(accountName)}/{Enc(groupName)}", cancellationToken);

    /// <summary>Returns the member Lists (and pinned versions) that a specific group_version searches.</summary>
    public Task<Response<GroupVersionMembersResponse>> GetVersionMembersAsync(string accountName, string groupName, int version, CancellationToken cancellationToken = default) =>
        _transport.GetAsync<GroupVersionMembersResponse>($"{Path}/versions/{Enc(accountName)}/{Enc(groupName)}/{version}", cancellationToken);

    /// <summary>Gets a Group's configuration, member Lists and IP whitelist.</summary>
    public Task<Response<GroupResponse>> GetAsync(string accountName, string groupName, CancellationToken cancellationToken = default) =>
        _transport.GetAsync<GroupResponse>($"{Path}/{Enc(accountName)}/{Enc(groupName)}", cancellationToken);

    /// <summary>Updates a Group's default search/suggest size and distance unit.</summary>
    public Task<Response<GroupResponse>> UpdateAsync(string accountName, string groupName, UpdateGroupRequest request, CancellationToken cancellationToken = default) =>
        _transport.PutAsync<GroupResponse>($"{Path}/{Enc(accountName)}/{Enc(groupName)}", request, cancellationToken);

    /// <summary>Adds a List (at a version) to a Group.</summary>
    public Task<Response<GroupResponse>> AddListAsync(string accountName, string groupName, AddListToGroupRequest request, CancellationToken cancellationToken = default) =>
        _transport.PutAsync<GroupResponse>($"{Path}/list/{Enc(accountName)}/{Enc(groupName)}", request, cancellationToken);

    /// <summary>Removes a List from a Group.</summary>
    public Task<Response<GroupResponse>> RemoveListAsync(string accountName, string groupName, string listName, CancellationToken cancellationToken = default) =>
        _transport.DeleteAsync<GroupResponse>($"{Path}/list/{Enc(accountName)}/{Enc(groupName)}/{Enc(listName)}", cancellationToken);

    /// <summary>Sets the embedding model for a Group; member Lists are re-embedded asynchronously.</summary>
    public Task<Response<GroupResponse>> SetModelAsync(string accountName, string groupName, SetGroupModelRequest request, CancellationToken cancellationToken = default) =>
        _transport.PutAsync<GroupResponse>($"{Path}/model/{Enc(accountName)}/{Enc(groupName)}", request, cancellationToken);

    /// <summary>Clears the embedding model for a Group, reverting it to keyword search.</summary>
    public Task<Response<GroupResponse>> RemoveModelAsync(string accountName, string groupName, CancellationToken cancellationToken = default) =>
        _transport.DeleteAsync<GroupResponse>($"{Path}/model/{Enc(accountName)}/{Enc(groupName)}", cancellationToken);

    /// <summary>Attaches a bring-your-own reranker to a Group.</summary>
    public Task<Response<GroupResponse>> SetRerankerAsync(string accountName, string groupName, SetGroupRerankerRequest request, CancellationToken cancellationToken = default) =>
        _transport.PostAsync<GroupResponse>($"{Path}/reranker/{Enc(accountName)}/{Enc(groupName)}", request, cancellationToken);

    /// <summary>Detaches the reranker from a Group, reverting to score-based ordering.</summary>
    public Task<Response<GroupResponse>> RemoveRerankerAsync(string accountName, string groupName, CancellationToken cancellationToken = default) =>
        _transport.DeleteAsync<GroupResponse>($"{Path}/reranker/{Enc(accountName)}/{Enc(groupName)}", cancellationToken);

    /// <summary>Moves a Group and its member Lists to a different subscription.</summary>
    public Task<Response<GroupResponse>> TransferAsync(string accountName, string groupName, TransferGroupRequest request, CancellationToken cancellationToken = default) =>
        _transport.PutAsync<GroupResponse>($"{Path}/transfer/{Enc(accountName)}/{Enc(groupName)}", request, cancellationToken);

    /// <summary>Removes an existing Group.</summary>
    public Task<Response> DeleteAsync(string accountName, string groupName, CancellationToken cancellationToken = default) =>
        _transport.DeleteAsync($"{Path}/{Enc(accountName)}/{Enc(groupName)}", cancellationToken);
}
